### ISSUE TITLE
Line number gutter width control

### ISSUE THREAD

**Issue by krstp** [zed-industries/zed#7334]
### Check for existing issues

- [X] Completed

### Describe the feature

I would like to be able to control the _line-number gutter_ `width`.

### If applicable, add mockups / screenshots to help present your vision of the feature

Now it is:
![Screenshot 2024-02-03 at 12 39 56 PM](https://github.com/zed-industries/zed/assets/4711498/b3dc2ce2-941c-4ef7-a318-271ac93ace40)
where I would like to be able to narrow it to something like this:
![Screenshot 2024-02-03 at 12 40 47 PM](https://github.com/zed-industries/zed/assets/4711498/85c5ee3d-90b0-4fb8-8f13-8260da27924a)

**Comment by diocletiann** [comment]
Option to turn off the folding arrow area would be nice as well.

**Comment by den-is** [comment]
The "line number" column is huuuuge!

**Comment by Caleb-T-Owens** [comment]
Hi! I've been trying out zed today and really enjoying it. The number column is however bonkers and is taking up 8% of my screen!

If a first step was to even just have an option to hide the line numbers, that would be great

**Comment by krstp** [comment]
@Caleb-T-Owens you can turn on/off line numbers with `CMD+;`

**Comment by Caleb-T-Owens** [comment]
Thanks @krstp and @SomeoneToIgnore ; Disabling code actions and folds seems like the best options for me 👍

**Comment by manaskarekar** [comment]
Another request to add adjustable gutter width. Thank you.

**Comment by galenelias** [comment]
The current logic is explicitly reserving 4 digits of line number space to avoid resizing the gutter as more lines are added.

This was done explicitly in https://github.com/zed-industries/zed/pull/4135.

It seems like either not reserving anything, or maybe just reserving two or three digits would be preferable for some folks.  Seems like the transition from 999 lines to 1000 is likely pretty rare, and some gutter size jump might be acceptable.

**Comment by brainrake** [comment]
This problem is way worse with proportional fonts (i know, niche), here's how it looks with code actions and folds disabled:
![screenshot-2024-09-20-19:42:06](https://github.com/user-attachments/assets/d3e1c5cd-5280-4595-a052-015f28709dfc)
Looks like a very wide character is used to determine the width, perhaps max width of digits would be more appropriate.

Disabling code actions doesn't change the width, it just hides the action icons.

Wouldn't it make sense to reserve as many digits as needed for the number of lines in the file? There would be no jumps and no wasted space.

In any case, a gutter width setting would be appreciated.

**Comment by galenelias** [comment]
> Looks like a very wide character is used to determine the width, perhaps max width of digits would be more appropriate.

It is using the letter 'm', which for a proportional font would likely be fair bit wider than the widest digit.


> Wouldn't it make sense to reserve as many digits as needed for the number of lines in the file? There would be no jumps and no wasted space.

I believe the choice to use such a large width is specifically to avoid the visual jumps when the file goes from 9 -> 10 lines, from 99 -> 100, and 999 -> 1000, etc.  Looks like VSCode has similar logic, although they only reserve 3 characters of width, not 4.

**Comment by uyen9vba** [comment]
Gutter width would be delightful. +1

**Comment by echostain** [comment]
> It seems like either not reserving anything, or maybe just reserving two or three digits would be preferable for some folks. 

Especially for those using relative line numbers

**Comment by shmarts** [comment]
+1

**Comment by matj1** [comment]
@brainrake
> Looks like a very wide character is used to determine the width, perhaps max width of digits would be more appropriate.

I found that the width of ‘m’ is used to get the width of a digit.  More information and an argument similar to yours is in [my comment to the mentioned pull request #4135](https://github.com/zed-industries/zed/pull/4135#issuecomment-2520697228).

**Comment by matj1** [comment]
@brainrake
I created an issue asking to change the character using the width of which the width of the gutter is calculated: #21860

**Comment by dkmar** [comment]
yeah we definitely need more config options for this lol. mine is insane

<img width="281" alt="Image" src="https://github.com/user-attachments/assets/44ab3f59-2a89-4ddd-bf81-e3a6a454e9f6" />

**Comment by playbahn** [comment]
I've found this to be a pretty okay workaround meanwhile. In `~/.config/zed/settings.json`:

    "gutter": {
        "line_numbers": false,
        "code_actions": false,
        "runnables": false,
        "folds": false
    }

If you actively use just your keyboard, `code_actions`, `runnables`, and `folds` are kinda unneeded. As for `line_numbers`, if you're not on `Vim mode`, it's just noise, and error messages while building/running always show line numbers along side the code at fault, and you can just `ctrl-g` to go to those lines.

**Comment by niksmac** [comment]
@playbahn That worked. I end up having the following 

```json
...
"gutter": {
    "line_numbers": true,
    "code_actions": false,
    "runnables": false,
    "folds": false
  }
```

**Comment by nickisnoble** [comment]
Here to also say, only like 5% of the files I work with get up to 1000+ lines, and the dead space is frustrating. Let us set that, or control the width.

**Comment by tommasoclini** [comment]
Maybe `line_numbers` could be a number, where 0 means no line numbers and other positive integers(up to 4 or 5 maybe) sets the default line numbers column width.

**Comment by thunderbolt004** [comment]
please add setting for gutter width

**Comment by brainrake** [comment]
#31959 does not solve the issue for proportional fonts, see https://github.com/zed-industries/zed/pull/31959#issuecomment-2962171687 
This issue should be reopened.

**Comment by MrSubidubi** [comment]
@brainrake This is a separate issue which is being tracked in #21860

**Comment by brainrake** [comment]
That's right, the issue with proportional fonts is tracked in https://github.com/zed-industries/zed/issues/21860. There is no need to reopen this one. Thank you for pointing it out.

**Comment by matj1** [comment]
Although it makes sense that this and #21860 are two different issues from a practical perspective, I am disappointed by that the solution to this likely does not work well with proportional fonts because of how this issue is stated. This issue is titled “Line number gutter width control”, not “Line number gutter width control unless a proportional font is used”.
